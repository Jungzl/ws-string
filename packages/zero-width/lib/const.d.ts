/**
 * Created by user on 2019/8/7.
 */
export declare const enum ENUM_ZERO_WIDTH {
    NON_JOINER = "\u200C",
    JOINER = "\u200D",
    SPACE = "\u200B",
    NO_BREAK_SPACE = "\uFEFF",
    BOM = "\uFEFF",
    LEFT_TO_RIGHT_MARK = "\u200E",
    RIGHT_TO_LEFT_MARK = "\u200F",
    WORD_JOINER = "\u2060",
    /**
     * no need remove
     * @see https://zh.wikipedia.org/wiki/%E7%B5%84%E5%90%88%E9%99%84%E5%8A%A0%E7%AC%A6%E8%99%9F
     * @see https://unicode-table.com/cn/034F/
     */
    COMBINING_GRAPHEME_JOINER = "\u034F"
}
/**
 * @see https://unicode-table.com/en/blocks/variation-selectors/
 */
export declare const enum ENUM_VARIATION_SELECTORS {
    /**
     * 	VARIATION SELECTOR-1
     */
    'FE00' = "\uFE00",
    /**
     * 	VARIATION SELECTOR-2
     */
    'FE01' = "\uFE01",
    /**
     * 	VARIATION SELECTOR-3
     */
    'FE02' = "\uFE02",
    /**
     * 	VARIATION SELECTOR-4
     */
    'FE03' = "\uFE03",
    /**
     * 	VARIATION SELECTOR-5
     */
    'FE04' = "\uFE04",
    /**
     * 	VARIATION SELECTOR-6
     */
    'FE05' = "\uFE05",
    /**
     * 	VARIATION SELECTOR-7
     */
    'FE06' = "\uFE06",
    /**
     * 	VARIATION SELECTOR-8
     */
    'FE07' = "\uFE07",
    /**
     * 	VARIATION SELECTOR-9
     */
    'FE08' = "\uFE08",
    /**
     * 	VARIATION SELECTOR-10
     */
    'FE09' = "\uFE09",
    /**
     * 	VARIATION SELECTOR-11
     */
    'FE0A' = "\uFE0A",
    /**
     * 	VARIATION SELECTOR-12
     */
    'FE0B' = "\uFE0B",
    /**
     * 	VARIATION SELECTOR-13
     */
    'FE0C' = "\uFE0C",
    /**
     * 	VARIATION SELECTOR-14
     */
    'FE0D' = "\uFE0D",
    /**
     * 	VARIATION SELECTOR-15
     */
    'FE0E' = "\uFE0E",
    /**
     * 	VARIATION SELECTOR-16
     */
    'FE0F' = "\uFE0F"
}
/**
 * too many give up
 * @see https://unicode-table.com/cn/blocks/variation-selectors-supplement/
 * @deprecated
 */
export declare const enum ENUM_VARIATION_SELECTORS_SUPPLEMENT {
}
export declare const enum ENUM_SPACE {
}
