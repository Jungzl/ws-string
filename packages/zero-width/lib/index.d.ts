export declare function existsZeroWidth(str: string): boolean;
export declare function removeVariationSelectors(str: string): string;
export declare function removeZeroWidth(str: string): string;
export declare function trimWithZeroWidth(str: string): string;
export declare function removeBom(str: string, unsafe?: boolean): string;
export default removeZeroWidth;
