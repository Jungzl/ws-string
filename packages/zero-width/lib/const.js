"use strict";
/**
 * Created by user on 2019/8/7.
 */
exports.__esModule = true;
