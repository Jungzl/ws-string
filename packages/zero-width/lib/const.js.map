{"version":3,"file":"const.js","sourceRoot":"","sources":["const.ts"],"names":[],"mappings":";AAAA;;GAEG;;AAEH,IAAkB,eAoBjB;AApBD,WAAkB,eAAe;IAEhC,wCAAqB,CAAA;IACrB,oCAAiB,CAAA;IACjB,mCAAgB,CAAA;IAEhB,4CAAyB,CAAA;IACzB,iCAAc,CAAA;IAEd,gDAA6B,CAAA;IAC7B,gDAA6B,CAAA;IAE7B,yCAAsB,CAAA;IAEtB;;;;OAIG;IACH,uDAAoC,CAAA;AACrC,CAAC,EApBiB,eAAe,GAAf,uBAAe,KAAf,uBAAe,QAoBhC;AAED;;GAEG;AACH,IAAkB,wBAkEjB;AAlED,WAAkB,wBAAwB;IAEzC;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;IACjB;;OAEG;IACH,2CAAiB,CAAA;AAClB,CAAC,EAlEiB,wBAAwB,GAAxB,gCAAwB,KAAxB,gCAAwB,QAkEzC;AAED;;;;GAIG;AACH,IAAkB,mCAGjB;AAHD,WAAkB,mCAAmC;AAGrD,CAAC,EAHiB,mCAAmC,GAAnC,2CAAmC,KAAnC,2CAAmC,QAGpD;AAED,IAAkB,UAGjB;AAHD,WAAkB,UAAU;AAG5B,CAAC,EAHiB,UAAU,GAAV,kBAAU,KAAV,kBAAU,QAG3B","sourcesContent":["/**\n * Created by user on 2019/8/7.\n */\n\nexport const enum ENUM_ZERO_WIDTH\n{\n\tNON_JOINER = '\\u200c',\n\tJOINER = '\\u200d',\n\tSPACE = '\\u200b',\n\n\tNO_BREAK_SPACE = '\\ufeff',\n\tBOM = '\\ufeff',\n\n\tLEFT_TO_RIGHT_MARK = '\\u200e',\n\tRIGHT_TO_LEFT_MARK = '\\u200f',\n\n\tWORD_JOINER = '\\u2060',\n\n\t/**\n\t * no need remove\n\t * @see https://zh.wikipedia.org/wiki/%E7%B5%84%E5%90%88%E9%99%84%E5%8A%A0%E7%AC%A6%E8%99%9F\n\t * @see https://unicode-table.com/cn/034F/\n\t */\n\tCOMBINING_GRAPHEME_JOINER = '\\u034f',\n}\n\n/**\n * @see https://unicode-table.com/en/blocks/variation-selectors/\n */\nexport const enum ENUM_VARIATION_SELECTORS\n{\n\t/**\n\t * \tVARIATION SELECTOR-1\n\t */\n\t'FE00' = '\\uFE00',\n\t/**\n\t * \tVARIATION SELECTOR-2\n\t */\n\t'FE01' = '\\uFE01',\n\t/**\n\t * \tVARIATION SELECTOR-3\n\t */\n\t'FE02' = '\\uFE02',\n\t/**\n\t * \tVARIATION SELECTOR-4\n\t */\n\t'FE03' = '\\uFE03',\n\t/**\n\t * \tVARIATION SELECTOR-5\n\t */\n\t'FE04' = '\\uFE04',\n\t/**\n\t * \tVARIATION SELECTOR-6\n\t */\n\t'FE05' = '\\uFE05',\n\t/**\n\t * \tVARIATION SELECTOR-7\n\t */\n\t'FE06' = '\\uFE06',\n\t/**\n\t * \tVARIATION SELECTOR-8\n\t */\n\t'FE07' = '\\uFE07',\n\t/**\n\t * \tVARIATION SELECTOR-9\n\t */\n\t'FE08' = '\\uFE08',\n\t/**\n\t * \tVARIATION SELECTOR-10\n\t */\n\t'FE09' = '\\uFE09',\n\t/**\n\t * \tVARIATION SELECTOR-11\n\t */\n\t'FE0A' = '\\uFE0A',\n\t/**\n\t * \tVARIATION SELECTOR-12\n\t */\n\t'FE0B' = '\\uFE0B',\n\t/**\n\t * \tVARIATION SELECTOR-13\n\t */\n\t'FE0C' = '\\uFE0C',\n\t/**\n\t * \tVARIATION SELECTOR-14\n\t */\n\t'FE0D' = '\\uFE0D',\n\t/**\n\t * \tVARIATION SELECTOR-15\n\t */\n\t'FE0E' = '\\uFE0E',\n\t/**\n\t * \tVARIATION SELECTOR-16\n\t */\n\t'FE0F' = '\\uFE0F',\n}\n\n/**\n * too many give up\n * @see https://unicode-table.com/cn/blocks/variation-selectors-supplement/\n * @deprecated\n */\nexport const enum ENUM_VARIATION_SELECTORS_SUPPLEMENT\n{\n\n}\n\nexport const enum ENUM_SPACE\n{\n\n}\n\n"]}