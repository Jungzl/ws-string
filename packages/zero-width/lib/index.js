"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Created by user on 2019/8/7.
 */
const re_1 = require("./re");
function existsZeroWidth(str) {
    return re_1.reVariationSelectorsAll.test(str);
}
exports.existsZeroWidth = existsZeroWidth;
function removeVariationSelectors(str) {
    return str.replace(re_1.reVariationSelectorsAll, '');
}
exports.removeVariationSelectors = removeVariationSelectors;
function removeZeroWidth(str) {
    return str
        .replace(re_1.reZeroWidthAll, '');
}
exports.removeZeroWidth = removeZeroWidth;
function trimWithZeroWidth(str) {
    return str
        .replace(re_1.reZeroWidthTrim, '');
}
exports.trimWithZeroWidth = trimWithZeroWidth;
function removeBom(str, unsafe) {
    if (unsafe) {
        return str.replace(re_1.reBomUnsafe, '');
    }
    return str.replace(re_1.reBomStrict, '');
}
exports.removeBom = removeBom;
exports.default = removeZeroWidth;
//# sourceMappingURL=index.js.map