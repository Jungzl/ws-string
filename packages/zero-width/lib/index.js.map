{"version":3,"file":"index.js","sourceRoot":"","sources":["index.ts"],"names":[],"mappings":";;AAAA;;GAEG;AACH,6BAAwH;AAExH,SAAgB,eAAe,CAAC,GAAW;IAE1C,OAAO,4BAAuB,CAAC,IAAI,CAAC,GAAG,CAAC,CAAA;AACzC,CAAC;AAHD,0CAGC;AAED,SAAgB,wBAAwB,CAAC,GAAW;IAEnD,OAAO,GAAG,CAAC,OAAO,CAAC,4BAAuB,EAAE,EAAE,CAAC,CAAA;AAChD,CAAC;AAHD,4DAGC;AAED,SAAgB,eAAe,CAAC,GAAW;IAE1C,OAAO,GAAG;SACR,OAAO,CAAC,mBAAc,EAAE,EAAE,CAAC,CAC5B;AACF,CAAC;AALD,0CAKC;AAED,SAAgB,iBAAiB,CAAC,GAAW;IAE5C,OAAO,GAAG;SACR,OAAO,CAAC,oBAAe,EAAE,EAAE,CAAC,CAC5B;AACH,CAAC;AALD,8CAKC;AAED,SAAgB,SAAS,CAAC,GAAW,EAAE,MAAgB;IAEtD,IAAI,MAAM,EACV;QACC,OAAO,GAAG,CAAC,OAAO,CAAC,gBAAW,EAAE,EAAE,CAAC,CAAA;KACnC;IAED,OAAO,GAAG,CAAC,OAAO,CAAC,gBAAW,EAAE,EAAE,CAAC,CAAA;AACpC,CAAC;AARD,8BAQC;AAED,kBAAe,eAAe,CAAA","sourcesContent":["/**\n * Created by user on 2019/8/7.\n */\nimport { _regexpMerge, reBomStrict, reBomUnsafe, reVariationSelectorsAll, reZeroWidthAll, reZeroWidthTrim } from './re';\n\nexport function existsZeroWidth(str: string)\n{\n\treturn reVariationSelectorsAll.test(str)\n}\n\nexport function removeVariationSelectors(str: string)\n{\n\treturn str.replace(reVariationSelectorsAll, '')\n}\n\nexport function removeZeroWidth(str: string)\n{\n\treturn str\n\t\t.replace(reZeroWidthAll, '')\n\t;\n}\n\nexport function trimWithZeroWidth(str: string)\n{\n\treturn str\n\t\t.replace(reZeroWidthTrim, '')\n\t\t;\n}\n\nexport function removeBom(str: string, unsafe?: boolean)\n{\n\tif (unsafe)\n\t{\n\t\treturn str.replace(reBomUnsafe, '')\n\t}\n\n\treturn str.replace(reBomStrict, '')\n}\n\nexport default removeZeroWidth\n"]}