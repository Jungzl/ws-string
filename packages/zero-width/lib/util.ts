/**
 * Created by user on 2019/8/12.
 */

/**
 * simple merge, not real regexp merge
 * @private
 */
export function _regexpMerge(re: (string | RegExp)[])
{
	return new RegExp('[' + re.map(r => (typeof r === 'string' ? r : r.source).replace(/^\[/, '').replace(/\]$/, ''))
		.join('') + ']', 'ug')
}
