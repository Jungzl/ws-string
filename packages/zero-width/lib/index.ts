import {
	reBomStrict,
	reBomUnsafe,
	reNBSP,
	reVariationSelectorsAll,
	reZeroWidthAll, reZeroWidthAllSafe,
	reZeroWidthTrim,
} from './re';
import { ENUM_SPACE } from './const';

export function existsZeroWidth(str: string)
{
	return reZeroWidthAll.test(str)
}

export function removeVariationSelectors(str: string)
{
	return str.replace(reVariationSelectorsAll, '')
}

export function removeZeroWidth(str: string, unsafe?: boolean)
{
	if (unsafe)
	{
		return str.replace(reZeroWidthAll, '')
	}

	return str
		.replace(reZeroWidthAllSafe, '')
	;
}

export function trimWithZeroWidth(str: string)
{
	return str
		.replace(reZeroWidthTrim, '')
		;
}

export function removeBom(str: string, unsafe?: boolean)
{
	if (unsafe)
	{
		return str.replace(reBomUnsafe, '')
	}

	return str.replace(reBomStrict, '')
}

export function nbspToSpace(str: string)
{
	return str.replace(reNBSP, ENUM_SPACE.SPACE)
}

export default removeZeroWidth
