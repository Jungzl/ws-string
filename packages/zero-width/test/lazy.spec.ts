import removeZeroWidth from '../lib';
import { toHex } from 'regexp-helper-core';
import { ENUM_ZERO_WIDTH } from '../lib/const';
import { inspect } from 'util';

describe(`allow VariationSelectors by default`, () =>
{

	[
		`[對熊]埃爾斯特西之森林Part6[要注意⚠️]`,
	].forEach(input =>
	{

		test(inspect(input), () =>
		{

			let expected = input;
			let actual = removeZeroWidth(expected);
			let actual2 = removeZeroWidth(expected, true);

			expect(actual).toStrictEqual(expected);
			expect(actual2).not.toStrictEqual(expected);
			//expect(actual).toBeInstanceOf(Date);
			expect(actual).toMatchSnapshot();
			expect(actual.length).toMatchSnapshot();
			expect(actual2).toMatchSnapshot();
			expect(actual2.length).toMatchSnapshot();

		});

	});

})

describe(`keep source`, () =>
{

	[
		'👩‍👩‍👧‍👦',
	].forEach(input =>
	{

		test(inspect(input), () =>
		{

			let expected = input;
			let actual = removeZeroWidth(expected);
			let actual2 = removeZeroWidth(expected, true);

			expect(actual).toStrictEqual(expected);
			expect(actual2).not.toStrictEqual(expected);
			//expect(actual).toBeInstanceOf(Date);
			expect(actual).toMatchSnapshot();
			expect(actual.length).toMatchSnapshot();
			expect(actual2).toMatchSnapshot();
			expect(actual2.length).toMatchSnapshot();
		});

	});

})

describe(`demo`, () =>
{

	[
		'\udb40\udd00\n\r\b\t\v\u00a0 　\u0009',
		`1\u20602`,
		toHex('\f'.codePointAt(0)),
		'\u0002',

		ENUM_ZERO_WIDTH.SPACE,
		ENUM_ZERO_WIDTH.NO_BREAK_SPACE,
		ENUM_ZERO_WIDTH.RIGHT_TO_LEFT_MARK,
		ENUM_ZERO_WIDTH.LEFT_TO_RIGHT_MARK,

	].forEach(input =>
	{

		test(inspect(input), () =>
		{

			let expected = input;
			let actual = removeZeroWidth(expected);
			let actual2 = removeZeroWidth(expected, true);

			expect(input.length).toMatchSnapshot();

			expect(actual).toMatchSnapshot();
			expect(actual.length).toMatchSnapshot();
			expect(actual2).toMatchSnapshot();
			expect(actual2.length).toMatchSnapshot();
		});

	});

})
