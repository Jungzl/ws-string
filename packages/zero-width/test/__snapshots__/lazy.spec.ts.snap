// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allow VariationSelectors by default [對熊]埃爾斯特西之森林Part6[要注意⚠️] 1`] = `"[對熊]埃爾斯特西之森林Part6[要注意⚠️]"`;

exports[`allow VariationSelectors by default [對熊]埃爾斯特西之森林Part6[要注意⚠️] 2`] = `24`;

exports[`allow VariationSelectors by default [對熊]埃爾斯特西之森林Part6[要注意⚠️] 3`] = `"[對熊]埃爾斯特西之森林Part6[要注意⚠]"`;

exports[`allow VariationSelectors by default [對熊]埃爾斯特西之森林Part6[要注意⚠️] 4`] = `23`;

exports[`demo  1`] = `1`;

exports[`demo  2`] = `""`;

exports[`demo  3`] = `1`;

exports[`demo  4`] = `""`;

exports[`demo  5`] = `1`;

exports[`demo 000c 1`] = `4`;

exports[`demo 000c 2`] = `"000c"`;

exports[`demo 000c 3`] = `4`;

exports[`demo 000c 4`] = `"000c"`;

exports[`demo 000c 5`] = `4`;

exports[`demo 1⁠2 1`] = `3`;

exports[`demo 1⁠2 2`] = `"12"`;

exports[`demo 1⁠2 3`] = `2`;

exports[`demo 1⁠2 4`] = `"12"`;

exports[`demo 1⁠2 5`] = `2`;

exports[`demo ​ 1`] = `1`;

exports[`demo ​ 2`] = `""`;

exports[`demo ​ 3`] = `0`;

exports[`demo ​ 4`] = `""`;

exports[`demo ​ 5`] = `0`;

exports[`demo ‎ 1`] = `1`;

exports[`demo ‎ 2`] = `""`;

exports[`demo ‎ 3`] = `0`;

exports[`demo ‎ 4`] = `""`;

exports[`demo ‎ 5`] = `0`;

exports[`demo ‏ 1`] = `1`;

exports[`demo ‏ 2`] = `""`;

exports[`demo ‏ 3`] = `0`;

exports[`demo ‏ 4`] = `""`;

exports[`demo ‏ 5`] = `0`;

exports[`demo 󠄀
	  　	 1`] = `11`;

exports[`demo 󠄀
	  　	 2`] = `
"󠄀

	  　	"
`;

exports[`demo 󠄀
	  　	 3`] = `11`;

exports[`demo 󠄀
	  　	 4`] = `
"

	  　	"
`;

exports[`demo 󠄀
	  　	 5`] = `9`;

exports[`demo ﻿ 1`] = `1`;

exports[`demo ﻿ 2`] = `""`;

exports[`demo ﻿ 3`] = `0`;

exports[`demo ﻿ 4`] = `""`;

exports[`demo ﻿ 5`] = `0`;

exports[`keep source 👩‍👩‍👧‍👦 1`] = `"👩‍👩‍👧‍👦"`;

exports[`keep source 👩‍👩‍👧‍👦 2`] = `11`;

exports[`keep source 👩‍👩‍👧‍👦 3`] = `"👩👩👧👦"`;

exports[`keep source 👩‍👩‍👧‍👦 4`] = `8`;
